### PR TITLE
Feature/add json formatter to lint python

### PR DESCRIPTION
**Pull request by catcombo**
Does anyone know why do we run `poetry run add-trailing-comma` two times? `inputs.exclude_trailing_comma_path` has a default value. So we can just remove the second call and remove `if: ${{ inputs.exclude_trailing_comma_path != '' }}` to handle both cases.

https://moneymeets.atlassian.net/browse/MD-6983

### CODE REVIEW

**Comment by catcombo** [comment]
Ok, I guess I know the reason why there are 2 commands. First command formats files in the `exclude_trailing_comma_path` without returning a non-zero exit code. Second command checks all files in all directories. I think we can replace 2 commands with one:
```bash
- run: find -path ${{ inputs.exclude_trailing_comma_path }} -prune -or -name '*.py' -print | xargs poetry run add-trailing-comma
  shell: bash
```
It works even if `exclude_trailing_comma_path` is empty. But we need to update `exclude_trailing_comma_path` values, they must starts with `./` (e.g. `exclude_trailing_comma_path: ./api/schema/`).

**Comment by mm-matthias** [comment]
@catcombo The original commands do three things:
1. Format all excluded files and do not complain about changes.
2. Format all normal files and complain about changes.
3. Run a git diff and verify there are no changes (neither in the excluded nor in the normal files).

With your proposed command will 3) still work? I have a little hard time to understand how you want to wrap the two commands in one, but then again I am no `find` expert.

**Comment by catcombo** [comment]
I'm also not a `find` expert, but I'm a google senior and the `man` doc reader :) I tested suggested command in the `insurance-api`. If you run `find -path ./api/schema -prune -or -name '*.py' -print | xargs poetry run add-trailing-comma` it won't complain about anything as expected. if you run `find -path -prune -or -name '*.py' -print | xargs poetry run add-trailing-comma` it will reformat a dozen files in `api/schema` and return a non-zero exit code. We don't run `git diff` here because `add-trailing-comma` will return a non-zero exit code in the case if it applied some changes to the files. I added `git diff` command to the PR only because JSON formatter command cannot return non-zero code if there are changes if the files.

**Review by zyv — APPROVED** [review]
I would leave the trailing comma commands as is I don't think the other way makes much difference, and you would have to update all the usages to a maybe less obvious form, so I don't believe that it's worth it.

**Review by marns93 — APPROVED** [review]
:+1: 

I would leave the commands as they are as well. I don't see a big benefit and it can break if somebody updates the path and removes the `./`.

If it is confusing for you, we should add a comment or add the `name` key with a meaningful description? WDYT?

**Comment by catcombo** [comment]
> If it is confusing for you, we should add a comment or add the `name` key with a meaningful description? WDYT?

It would be great! But I'm also fine if you keep it like it is. Maybe it's just me so slow to figure out how the action it works :)

I have only one question regarding calling composite action (`check-git-diff`) from another composite action. Is this allowed?

**Comment by marns93** [comment]
> I have only one question regarding calling composite action (`check-git-diff`) from another composite action. Is this allowed?

This should be possible.

**Comment by catcombo** [comment]
@marns93 Thanks! Should we wait for @felix11h to review the changes before merging the PR?

**Comment by felix11h** [comment]
> I have only one question regarding calling composite action (`check-git-diff`) from another composite action. Is this allowed?

I wasn't sure myself and had the same question. If this works, I'm happy with the way it is done here.

I just want to comment on the bigger picture. We are now validating JSON formats in all of our Python repositories. It's reasonable to assume that checks will fail in some repositories and that we have to change formatting accordingly. I'm not sure if there could possibly be cases in which we don't want to change the formatting.

The point that I'm trying to make is that this is going to create some work, and it would be important that all developers are aware of the change. For me personally, with changes like this, it can be difficult to understand or remember the reason why we are introducing this. I understand that this came up in a discussion in some other pull request, but if I had to explain to a new developer why we are doing this, I wouldn't be able to say. Is this something we can improve?

**Review by felix11h — APPROVED** [review]

**Comment by zyv** [comment]
> I understand that this came up in a discussion in some other pull request, but if I had to explain to a new developer why we are doing this, I wouldn't be able to say. Is this something we can improve?

Oh, that's the easiest thing of all, if you already have a reference that you can accept. See like we have `prettier` for JavaScript code and `black` for Python code? Now we've finally got to introducing the same for `JSON`.

If you don't have a reference that you can accept, then it's more difficult, but the reasons are always the same:

1. If one person is working consistently -> OK
2. If one person is unable to work consistently but doesn't see a problem with that -> OK
3. If several persons are working together and are able to remain consistent -> OK
4. If several persons are working together, some are unable to work consistently, and some have a problem with that -> conflict

So force-linting and auto-fixing is added to resolve (4) when the pain gets too high, and even for consistent people it's a welcome help that supports them and allows to put their concentration to better uses.

In one of the PRs where I had to collaborate the pain for JSON became too high, so this PR was created. Of course it will cause work, at least initially, but this was intended.

If this is what you mean, and it helps, then feel free to add it to wherever the spoken guidelines are located.

**Comment by catcombo** [comment]
> I just want to comment on the bigger picture. We are now validating JSON formats in all of our Python repositories. It's reasonable to assume that checks will fail in some repositories and that we have to change formatting accordingly. I'm not sure if there could possibly be cases in which we don't want to change the formatting.

This is a good point. I've checked the available JSON files in our Python projects and found files created by external tools that I wouldn't touch:
```
moneymeets-tenants/signature-pad/package-lock.json
moneymeets-tenants/signature-pad/package.json
moneymeets-tenants/signature-pad/tsconfig.json
moneymeets-tenants/mjml/package-lock.json
moneymeets-tenants/mjml/package.json
fondsnet-sdk/puppeteer/package-lock.json
fondsnet-sdk/puppeteer/package.json
```
But fortunately, their content remain the same even after the formatting command is applied.

**Comment by felix11h** [comment]
> If this is what you mean, and it helps, then feel free to add it to wherever the spoken guidelines are located.

This does help, thank you, @zyv. I think what I was mostly missing was a ticket giving context for this change, but I just saw that we actually do have a ticket for this: https://moneymeets.atlassian.net/browse/MD-6983. 

> If this is what you mean, and it helps, then feel free to add it to wherever the spoken guidelines are located.

I have made a link to the pull request here in the ticket comments, I believe this is sufficient. As you correctly say, this is not too different from other formatting tools (black, prettier) that we are already using.

**Comment by felix11h** [comment]
> This is a good point. I've checked the available JSON files in our Python projects and found files created by external tools that I wouldn't touch:

Thank you for checking this! Do you then believe it's necessary to exclude those files from checking them, or would one have to run the formatter after updating the npm dependencies?

**Comment by zyv** [comment]
> > This is a good point. I've checked the available JSON files in our Python projects and found files created by external tools that I wouldn't touch:
> 
> Thank you for checking this! Do you then believe it's necessary to exclude those files from checking them, or would one have to run the formatter after updating the npm dependencies?

We could exclude them, but they are written by tools and are already consistent, so excluding or including them doesn't make a difference in practice. One could even argue that if they are inconsistent it's a bug that has to be reported to the tool and fixed there...

**Comment by catcombo** [comment]
> > This is a good point. I've checked the available JSON files in our Python projects and found files created by external tools that I wouldn't touch:
> 
> Thank you for checking this! Do you then believe it's necessary to exclude those files from checking them, or would one have to run the formatter after updating the npm dependencies?

I agree with @zyv Lets don't introduce exclude patterns until it's really needed.

**Comment by felix11h** [comment]
> so excluding or including them doesn't make a difference in practice

I quickly tried this in moneymeets-tenants. It seems that package-lock.json formatting is consistent with the json.tool target format, there is no diff after running json.tool on these files.  So, you are right, exclusion/inclusion should not make a difference in practice 👍

**Pull request merged**
